### PR TITLE
Add reboot test for FLUO

### PR DESCRIPTION
**Pull request by surajssd**

### CODE REVIEW

File: test/components/flatcar-linux-update-operator/fluo_reboot_test.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build packet_fluo
+// +build disruptivee2e
+
+package fluo //nolint:testpackage
```
Comment (invidian) [inline test/components/flatcar-linux-update-operator/fluo_reboot_test.go:18]: Why not `fluo_test` ?

File: test/components/flatcar-linux-update-operator/fluo_reboot_test.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+	if len(nodes) < 1 {
+		t.Fatalf("Wanted one or more nodes found none.")
+	}
+
+	// Select a node to remove annotation
+	chosenNode := nodes[0]
+	t.Logf("Chosen node to reboot: %s", chosenNode.Name)
+
+	// Remove annotation.
+	delete(chosenNode.Annotations, "flatcar-linux-update.v1.flatcar-linux.net/reboot-paused")
```
Comment (invidian) [inline test/components/flatcar-linux-update-operator/fluo_reboot_test.go:54]: This assumes the annotation is in place. Can we just not set it when we run `packet_fluo` pipeline?

Or should we rather say in the comment that we ensure that the annotation is not set?

File: test/components/flatcar-linux-update-operator/fluo_reboot_test.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+	if _, err := client.CoreV1().Nodes().Update(context.Background(), &chosenNode, metav1.UpdateOptions{}); err != nil {
+		t.Fatalf("Removing node annotation 'flatcar-linux-update.v1.flatcar-linux.net/reboot-paused': %v", err)
+	}
+
+	// Wait for the node to reboot and keep checking if the following annotation is added:
```
Comment (invidian) [inline test/components/flatcar-linux-update-operator/fluo_reboot_test.go:60]: Why would node reboot itself if it's up to date? We should assume Packet may serve both up to date and not up to date images. Otherwise test will be flaky. We also don't explicitly set installed version so we can control which version gets installed.

File: test/components/flatcar-linux-update-operator/fluo_reboot_test.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+	// "flatcar-linux-update.v1.flatcar-linux.net/status": "UPDATE_STATUS_IDLE"
```
Comment (invidian) [inline test/components/flatcar-linux-update-operator/fluo_reboot_test.go:61]: What this annotation indicate? We don't explain it here.

**Review by invidian — CHANGES_REQUESTED** [review]

File: test/components/flatcar-linux-update-operator/fluo_reboot_test.go (hunk 1 above)
Comment (surajssd) [reply test/components/flatcar-linux-update-operator/fluo_reboot_test.go:18]: There will be two packages in the same directory and I am not sure if it is a good thing or a bad thing.

Comment (surajssd) [reply test/components/flatcar-linux-update-operator/fluo_reboot_test.go:18]: Also the package would be `github.com/kinvolk/lokomotive/test/components/flatcar-linux-update-operator/fluo_test` which is test and again test.

File: test/components/flatcar-linux-update-operator/fluo_reboot_test.go (hunk 2 above)
Comment (surajssd) [reply test/components/flatcar-linux-update-operator/fluo_reboot_test.go:54]: I don't want to reboot all the nodes in the pipeline randomly. So I will let it happen in the controlled way. I will just add a comment.

File: test/components/flatcar-linux-update-operator/fluo_reboot_test.go (hunk 3 above)
Comment (surajssd) [reply test/components/flatcar-linux-update-operator/fluo_reboot_test.go:60]: Changed the comment.

File: test/components/flatcar-linux-update-operator/fluo_reboot_test.go (hunk 4 above)
Comment (surajssd) [reply test/components/flatcar-linux-update-operator/fluo_reboot_test.go:61]: Added the comment.

File: test/components/flatcar-linux-update-operator/fluo_reboot_test.go (hunk 1 above)
Comment (invidian) [reply test/components/flatcar-linux-update-operator/fluo_reboot_test.go:18]: I don't think it's a problem. Package name doesn't really matter here, but we should avoid adding `nolint` statements where possible.

**Comment by surajssd** [comment]
I am not sure what's going wrong in the 'Packet FLUO' pipeline it always errors out on

```
=== CONT  TestReboot
    util.go:81: using KUBECONFIG=/root/lokoctl-assets/cluster-assets/auth/kubeconfig
    fluo_reboot_test.go:52: Chosen node to reboot: ci1604058269-re-general-worker-0
    fluo_reboot_test.go:61: Annotation "flatcar-linux-update.v1.flatcar-linux.net/reboot-paused" not found on the node.
--- FAIL: TestReboot (0.03s)
```

**Comment by surajssd** [comment]
Oh CI code only applies the annotations to packet and to packet_fluo :facepalm:

File: test/components/flatcar-linux-update-operator/fluo_reboot_test.go [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+	if err != nil {
+		t.Fatalf("Listing nodes with label 'fluo-test-pool=true': %v", err)
+	}
+
+	nodes := nodesList.Items
+	if len(nodes) < 1 {
+		t.Fatalf("Wanted one or more nodes with label 'fluo-test-pool=true', found none.")
+	}
+
+	// Select a node to remove annotation
```
Comment (invidian) [inline test/components/flatcar-linux-update-operator/fluo_reboot_test.go:50]:
```suggestion
	// Select a node to remove annotation.
```

**Review by invidian — COMMENTED** [review]
I don't fully understand what this is testing exactly. There seems to be nothing specific to Lokomotive in those tests, perhaps they should be moved to FLUO repository?

Also, how possible reboots play with #1084? Won't reboot interrupt the other test and make it fail?

**Comment by surajssd** [comment]
> I don't fully understand what this is testing exactly.

This tests FLUO on lokomotive and if it can safely reboot machine. This is most basic test for stateless nodes. This pipeline also runs rook-ceph tests and hence we can extend the testing on FLUO further to test upgrading nodes with rook-ceph running.

> There seems to be nothing specific to Lokomotive in those tests, perhaps they should be moved to FLUO repository?

The argument is true for so many other component e2e tests like prometheus-operator, metallb, contour, etc. 

> Also, how possible reboots play with #1084? Won't reboot interrupt the other test and make it fail?

This is marked as disruptive e2e test.

File: test/components/flatcar-linux-update-operator/fluo_reboot_test.go [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+		}
+
+		// Test passed since the annotation is added by the FLUO to the nodes.
+		if val == statusAnnotationVal {
+			return true, nil
+		}
+
+		return false, nil
+	}); err != nil {
+		t.Fatalf("Waiting for the node to add annotation '%s:%s' %v",
```
Comment (knrt10) [inline test/components/flatcar-linux-update-operator/fluo_reboot_test.go:95]: Should this say wait? As the test is failed. I would say failed to add the annotation. What do you think?

Comment (surajssd) [reply test/components/flatcar-linux-update-operator/fluo_reboot_test.go:95]: I think the idea was to say what we were doing when this failure happened, isn't it?

Comment (knrt10) [reply test/components/flatcar-linux-update-operator/fluo_reboot_test.go:95]: Cool.

File: test/components/flatcar-linux-update-operator/fluo_reboot_test.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+	if _, err := client.CoreV1().Nodes().Update(context.Background(), &chosenNode, metav1.UpdateOptions{}); err != nil {
+		t.Fatalf("Removing node annotation %q: %v", annotation, err)
+	}
+
+	// Wait for the FLUO to add following annotation key value pair to the chosen node. This may
+	// involve node reboot if the OS is outdated.
+	// "flatcar-linux-update.v1.flatcar-linux.net/status": "UPDATE_STATUS_IDLE"
+	statusAnnotationKey := "flatcar-linux-update.v1.flatcar-linux.net/status"
+	statusAnnotationVal := "UPDATE_STATUS_IDLE"
```
Comment (invidian) [inline test/components/flatcar-linux-update-operator/fluo_reboot_test.go:76]: >This tests FLUO on lokomotive and if it can safely reboot machine. This is most basic test for stateless nodes. This pipeline also runs rook-ceph tests and hence we can extend the testing on FLUO further to test upgrading nodes with rook-ceph running.

So, as far as I understand, what we _want_ to test is `FLUO on Lokomotive can safely reboot machines`. How do we define safely? What are our safety checks here?

From what I understand here, we test: _FLUO on Lokomotive runs update agent which can read `update_engine` status and put it into Node annotations._

In this process, an operator is not really involved (except labeling nodes with label, which allows agent DaemonSet to be scheduled on those nodes).

We also test, that node can reach `UPDATE_STATUS_IDLE` status, but rebooting is not deterministic.

To summarize, I think name of the test as it is now (`TestReboot`) should be changed as it is misleading at the moment.

 Perhaps we can rename it to something like `TestNodeCanReachIdleState` to better represent what are we testing?

**Review by invidian — CHANGES_REQUESTED** [review]

File: test/components/flatcar-linux-update-operator/fluo_reboot_test.go (hunk 7 above)
Comment (surajssd) [reply test/components/flatcar-linux-update-operator/fluo_reboot_test.go:76]: done.

**Review by invidian — APPROVED** [review]
LGTM

**Review by knrt10 — APPROVED** [review]

**Review by iaguis — APPROVED** [review]
